### PR TITLE
Integration: Date Picker [LG-3152]

### PR DESCRIPTION
**Pull request by TheSonOfThomp**
## ✍️ Proposed changes

Creates the Date Picker component

RFC Document: https://docs.google.com/document/d/1HussHpUyylDvQFNlwadxGj8A8ThpO3Uzj_D6HoYY7LQ/

🎟 _Jira ticket:_ https://jira.mongodb.org/browse/LG-3152

### For bug fixes, new features & breaking changes

- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] I have run `yarn changeset` and documented my changes

### For new components

- [ ] I have added my new package to the global tsconfig
- [ ] I have added my new package to the Table of Contents on the global README
- [ ] I have verified the Live Example looks as intended on the design website.

### CODE REVIEW

File: packages/date-picker/src/shared/constants.ts [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+  'July',
+  'August',
+  'September',
+  'October',
+  'November',
+  'December',
+];
+
+/** The default earliest selectable date */
+export const MIN_DATE = new Date(Date.UTC(1970, Month.January, 1));
```
Comment (shaneeza) [inline packages/date-picker/src/shared/constants.ts:42]: Where do these min//max dates come from?

File: packages/date-picker/README.md [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+```shell
+yarn add @leafygreen-ui/date-picker
+```
+
+### NPM
+
+```shell
+npm install @leafygreen-ui/date-picker
+```
```
Comment (bruugey) [inline packages/date-picker/README.md:19]: need docs

File: packages/date-picker/rollup.config.mjs [hunk 3]
```diff
@@ -0,0 +1,14 @@
+import {
+  esmConfig,
+  storiesConfig,
+  umdConfig,
+} from '@lg-tools/build/config/rollup.config.mjs';
+
+const sharedConfig = [esmConfig, umdConfig].map(config => ({
```
Comment (bruugey) [inline packages/date-picker/rollup.config.mjs:7]: are we still using this?

File: packages/date-picker/src/DatePicker/DatePicker.testutils.tsx [hunk 4]
```diff
@@ -0,0 +1,213 @@
+import React from 'react';
+import {
+  fireEvent,
+  // prettyDOM,
```
Comment (bruugey) [inline packages/date-picker/src/DatePicker/DatePicker.testutils.tsx:4]: rm

File: packages/date-picker/src/DatePicker/DatePickerComponent/DatePickerComponent.tsx [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+export const DatePickerComponent = forwardRef<
+  HTMLDivElement,
+  DatePickerComponentProps
+>(({ ...rest }: DatePickerComponentProps, fwdRef) => {
+  const { isOpen, menuId, disabled, isSelectOpen } = useDatePickerContext();
+  const {
+    refs,
+    value,
+    closeMenu,
```
Comment (bruugey) [inline packages/date-picker/src/DatePicker/DatePickerComponent/DatePickerComponent.tsx:28]: should closeMenu and isOpen be colocated in the same context?

Comment (TheSonOfThomp) [reply packages/date-picker/src/DatePicker/DatePickerComponent/DatePickerComponent.tsx:28]: Ideally yes, but the `openMenu` and `closeMenu` functions trigger side effects that will be different between single & range picker, so we can't include them in the shared `useDatePickerContext`. 

If this is something we want to resolve we can either:
1. Move the `isOpen` state from `DatePickerContext` to `SingleDateContext`. (This will mean re-creating this state in `DateRangeContext` as well)
2. Rename the current state var to `_isOpen` and export `isOpen` from `DatePickerContext`

Neither are hard to do, this is just how the contexts ended up working

File: packages/date-picker/src/DatePicker/DatePicker.spec.tsx [hunk 6]
```diff
@@ -0,0 +1,1353 @@
...
+      });
+
+      test('formField contains label & input elements', () => {
+        const { getByTestId, getByRole } = render(
+          <DatePicker label="Label" data-testid="lg-date-picker" />,
+        );
+        const formField = getByTestId('lg-date-picker');
+        const inputContainer = getByRole('combobox');
+        expect(formField.querySelector('label')).toBeInTheDocument();
+        expect(formField.querySelector('label')).toHaveTextContent('Label');
```
[the remark this answers is not in the document]
Comment (TheSonOfThomp) [reply packages/date-picker/src/DatePicker/DatePicker.spec.tsx:70]: We should 👍

File: packages/date-picker/src/DatePicker/DatePickerComponent/DatePickerComponent.tsx (hunk 5 above)
Comment (bruugey) [reply packages/date-picker/src/DatePicker/DatePickerComponent/DatePickerComponent.tsx:28]: Ah interesting. I'd lean towards option 1, personally

File: packages/date-picker/src/DatePicker/DatePicker.types.ts [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+  /** The initial selected date. Ignored if `value` is provided */
+  initialValue?: DateType;
+
+  /**
+   * A callback fired when validation should run, based on our [form validation guidelines](https://www.mongodb.design/foundation/forms/#form-validation-error-handling).
+   * Use this callback to compute the correct `state` value.
+   *
+   * Callback date argument will be a Date object in ISO 8601 format, and in UTC time.
+   */
+  handleValidation?: (value?: DateType) => void;
```
[the remark this answers is not in the document]
Comment (bruugey) [reply packages/date-picker/src/DatePicker/DatePicker.types.ts:30]: kk lgtm.

File: packages/date-picker/README.md (hunk 2 above)
Comment (TheSonOfThomp) [reply packages/date-picker/README.md:19]: We need to decide how we want to handle README documentation going forward. What should go here vs TSDoc & Storybook?

File: packages/date-picker/src/DatePicker/DatePickerInput/DatePickerInput.tsx [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+      setValue,
+      openMenu,
+      toggleMenu,
+      handleValidation,
+    } = useSingleDateContext();
+
+    /** Called when the input's Date value has changed */
+    const handleInputValueChange = (inputVal?: Date | null) => {
+      if (!isSameUTCDay(inputVal, value)) {
+        handleValidation?.(inputVal);
```
Comment (bruugey) [inline packages/date-picker/src/DatePicker/DatePickerInput/DatePickerInput.tsx:52]: should we call onSegmentChange here?

File: packages/date-picker/src/DatePicker/DatePicker.spec.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (TheSonOfThomp) [reply packages/date-picker/src/DatePicker/DatePicker.spec.tsx:70]: We can do this in one line with [Custom Matchers](https://github.com/mongodb/leafygreen-ui/pull/2046) 😉

File: packages/date-picker/src/DatePicker/DatePickerInput/DatePickerInput.tsx (hunk 8 above)
Comment (TheSonOfThomp) [reply packages/date-picker/src/DatePicker/DatePickerInput/DatePickerInput.tsx:52]: Unless there's a use case you're seeing where there's a bug, the segment change handler should already be called within `DateInputBox`

File: packages/date-picker/src/shared/components/DateInput/DateInputBox/DateInputBox.tsx [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+ *
+ * Provided value is assumed to be UTC.
+ *
+ * Argument passed into `setValue` callback is also in UTC
+ * @internal
+ */
+export const DateInputBox = React.forwardRef<HTMLDivElement, DateInputBoxProps>(
+  (
+    {
+      value: dateValue,
```
Comment (bruugey) [inline packages/date-picker/src/shared/components/DateInput/DateInputBox/DateInputBox.tsx:46]: nit but these are internal only components - if you wanted to change the prop names so you didn't have to alias them, you could

File: packages/date-picker/src/shared/components/DateInput/DateInputBox/DateInputBox.tsx [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+      prevSegments?: DateSegmentsState,
+    ) => {
+      const hasAnySegmentChanged = !isEqual(newSegments, prevSegments);
+
+      if (hasAnySegmentChanged) {
+        const utcDate = newDateFromSegments(newSegments);
+        const areAllSegmentsEmpty = !doesSomeSegmentExist(newSegments);
+
+        if (utcDate) {
+          // Only update the value iff all parts are set, and create a valid date.
```
Comment (bruugey) [inline packages/date-picker/src/shared/components/DateInput/DateInputBox/DateInputBox.tsx:77]: nit: `if`

File: packages/date-picker/src/shared/components/DateInput/DateInputBox/DateInputBox.tsx [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+    const { theme } = useDarkMode();
+
+    const containerRef = useForwardedRef(fwdRef, null);
+
+    /**
+     * When a segment is updated,
+     * trigger a `change` event for the segment, and
+     * update the external Date value if necessary
+     */
+    const onSegmentsUpdate = (
```
[the remark this answers is not in the document]
Comment (TheSonOfThomp) [reply packages/date-picker/src/shared/components/DateInput/DateInputBox/DateInputBox.tsx:66]: the `onSegmentChange` handler (i.e. the `onChange` handler passed into the DatePicker component) should be called whenever the value of the segment input element changes. The Segment component handles most cases of this, so we only need to manually call it in the `handleSegmentChange` callback (since we want to perform side effects)

File: packages/date-picker/src/shared/components/DateInput/DateInputBox/DateInputBox.spec.tsx [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+    describe('renders segments in the correct order', () => {
+      test('iso8601', () => {
+        const result = renderDateInputBox(undefined, { dateFormat: 'iso8601' });
+        const segments = result.getAllByRole('spinbutton');
+        expect(segments[0]).toHaveAttribute('aria-label', 'year');
+        expect(segments[1]).toHaveAttribute('aria-label', 'month');
+        expect(segments[2]).toHaveAttribute('aria-label', 'day');
+      });
+
+      test('en-US', () => {
```
Comment (bruugey) [inline packages/date-picker/src/shared/components/DateInput/DateInputBox/DateInputBox.spec.tsx:87]: should we test for every region that we support?

File: packages/date-picker/src/shared/components/DateInput/DateInputBox/DateInputBox.spec.tsx [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+        { dateFormat: 'iso8601' },
+      );
+      userEvent.click(yearInput);
+      userEvent.tab();
+      expect(monthInput).toHaveFocus();
+      userEvent.tab();
+      expect(dayInput).toHaveFocus();
+    });
+
+    // Arrow key interaction tested in DateInputSegment
```
Comment (bruugey) [inline packages/date-picker/src/shared/components/DateInput/DateInputBox/DateInputBox.spec.tsx:227]: TODO

File: packages/date-picker/src/shared/components/DateInput/DateInputSegment/DateInputSegment.tsx [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+    const maxLength = getSegmentMaxLength(segment);
+
+    /** Prevent non-numeric values from triggering a change event */
+    const handleChange: ChangeEventHandler<HTMLInputElement> = e => {
+      const { target } = e;
+      const numericValue = Number(target.value);
+
+      if (!isNaN(numericValue)) {
+        const stringValue = numericValue.toString();
+        onChange?.({
```
Comment (bruugey) [inline packages/date-picker/src/shared/components/DateInput/DateInputSegment/DateInputSegment.tsx:70]: should this onChange be a required prop?

File: packages/date-picker/src/shared/components/DateInput/DateInputSegment/DateInputSegment.tsx [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+    // Number inputs also don't support the `selectionStart`/`End` API
+    return (
+      <input
+        {...rest}
+        aria-label={segment}
+        id={segment}
+        ref={inputRef}
+        type="text"
+        pattern={pattern}
+        role="spinbutton"
```
Comment (bruugey) [inline packages/date-picker/src/shared/components/DateInput/DateInputSegment/DateInputSegment.tsx:129]: TIL

File: packages/date-picker/src/DatePicker/DatePickerMenu/DatePickerMenu.tsx [hunk 16]
```diff
@@ -0,0 +1,263 @@
...
+      getHighlightedCell,
+    } = useSingleDateContext();
+
+    const ref = useForwardedRef(fwdRef, null);
+    const cellRefs = refs.calendarCellRefs;
+    const headerRef = useRef<HTMLDivElement>(null);
+    const calendarRef = useRef<HTMLTableElement>(null);
+
+    const prevValue = usePrevious(value);
+    // const prevOpen = usePrevious(isOpen);
```
Comment (bruugey) [inline packages/date-picker/src/DatePicker/DatePickerMenu/DatePickerMenu.tsx:65]: rm

File: packages/date-picker/src/DatePicker/DatePickerMenu/DatePickerMenu.stories.tsx [hunk 17]
```diff
@@ -0,0 +1,235 @@
+/* eslint-disable react-hooks/rules-of-hooks, react/prop-types */
+import React, { useRef, useState } from 'react';
+import { Decorator, StoryFn, StoryObj } from '@storybook/react';
+import { userEvent, within } from '@storybook/testing-library';
+import { last, omit } from 'lodash';
```
Comment (bruugey) [inline packages/date-picker/src/DatePicker/DatePickerMenu/DatePickerMenu.stories.tsx:5]: ditto here

File: packages/date-picker/src/DatePicker/DatePickerMenu/DatePickerMenuHeader/utils/getMonthOptions/index.ts [hunk 18]
```diff
@@ -0,0 +1,39 @@
+import { isNull } from 'lodash';
```
Comment (bruugey) [inline packages/date-picker/src/DatePicker/DatePickerMenu/DatePickerMenuHeader/utils/getMonthOptions/index.ts:1]: ditto here

File: packages/date-picker/src/shared/components/DateInput/DateInputSegment/DateInputSegment.spec.tsx [hunk 19]
```diff
@@ -0,0 +1,262 @@
+import React from 'react';
+import { jest } from '@jest/globals';
+import { render } from '@testing-library/react';
+import userEvent from '@testing-library/user-event';
+import { mean, round } from 'lodash';
```
Comment (bruugey) [inline packages/date-picker/src/shared/components/DateInput/DateInputSegment/DateInputSegment.spec.tsx:5]: ditto here

File: packages/date-picker/src/shared/components/DateInput/DateInputBox/DateInputBox.tsx (hunk 10 above)
Comment (TheSonOfThomp) [reply packages/date-picker/src/shared/components/DateInput/DateInputBox/DateInputBox.tsx:77]: "iff" is correct. [If and only if](https://en.wikipedia.org/wiki/If_and_only_if)

Update the value _if_ all parts are set & valid
 - does not exclude the possibility of other cases where the value can be updated
Update the value _only if_ all parts are set & valid
 - does not exclude the possibility of _not_ updating the value when all parts are set & valid

File: packages/date-picker/src/DatePicker/DatePickerMenu/DatePickerMenu.tsx [hunk 20]
```diff
@@ -0,0 +1,263 @@
...
+        handleCalendarCellClick(day);
+      }
+    };
+
+    // Focus trap
+    const handleMenuKeyPress: KeyboardEventHandler<HTMLDivElement> = e => {
+      const { key } = e;
+
+      // Implementing custom focus-trap logic,
+      // since focus-trap-react focuses the first element immediately on mount
```
Comment (bruugey) [inline packages/date-picker/src/DatePicker/DatePickerMenu/DatePickerMenu.tsx:149]: non-blocking but we should consider abstracting and moving this to lib

File: packages/date-picker/src/shared/components/DateInput/DateInputBox/DateInputBox.spec.tsx (hunk 12 above)
Comment (TheSonOfThomp) [reply packages/date-picker/src/shared/components/DateInput/DateInputBox/DateInputBox.spec.tsx:87]: We're testing the one's we've explicitly stated we're supporting. All others _should_ work, but there may be some small bugs here-&-there

File: packages/date-picker/src/shared/components/DateInput/DateInputSegment/DateInputSegment.tsx (hunk 14 above)
Comment (TheSonOfThomp) [reply packages/date-picker/src/shared/components/DateInput/DateInputSegment/DateInputSegment.tsx:70]: It _could_ be a required prop into `DateInputSegment`, since it will always be provided by `DateInputBox`.

I left it as optional, since `onChange` is not a required prop on `DatePicker` itself

File: packages/date-picker/src/DatePicker/DatePickerMenu/DatePickerMenu.tsx [hunk 21]
```diff
@@ -0,0 +1,263 @@
...
+  menuCalendarGridStyles,
+  menuContentStyles,
+  menuWrapperStyles,
+} from './DatePickerMenu.styles';
+import { DatePickerMenuProps } from './DatePickerMenu.types';
+import { DatePickerMenuHeader } from './DatePickerMenuHeader';
+
+export const DatePickerMenu = forwardRef<HTMLDivElement, DatePickerMenuProps>(
+  ({ onKeyDown, ...rest }: DatePickerMenuProps, fwdRef) => {
+    const today = useMemo(() => setToUTCMidnight(new Date(Date.now())), []);
```
[the remark this answers is not in the document]
Comment (TheSonOfThomp) [reply packages/date-picker/src/DatePicker/DatePickerMenu/DatePickerMenu.tsx:43]: Since we don't care about time zone, it doesn't matter if "now" is 11:59pm or 12:01am, "today" is the same date regardless

File: packages/date-picker/src/shared/components/DateInput/DateFormField/DateFormField.spec.tsx [hunk 22]
```diff
@@ -0,0 +1,8 @@
+import React from 'react';
+import { render } from '@testing-library/react';
+
+import { DateFormField } from '.';
+
+describe('packages/date-picker/shared/date-form-field', () => {
+  test('renders', () => {});
```
Comment (bruugey) [inline packages/date-picker/src/shared/components/DateInput/DateFormField/DateFormField.spec.tsx:7]: TODO

File: packages/date-picker/src/DatePicker/DatePickerInput/DatePickerInput.stories.tsx [hunk 23]
```diff
@@ -0,0 +1,92 @@
+/* eslint-disable react/prop-types */
+import React, { useEffect, useState } from 'react';
+import { StoryFn } from '@storybook/react';
+import { isValid } from 'date-fns';
+import { isUndefined } from 'lodash';
```
[the remark this answers is not in the document]
Comment (TheSonOfThomp) [reply packages/date-picker/src/DatePicker/DatePickerInput/DatePickerInput.stories.tsx:5]: This technically shouldn't matter for `.stories` or `.spec` files since these are not included in the component file

**Comment by bruugey** [comment]
I also can't tab from the arrow/controlers into the CalendarGrid

File: packages/date-picker/src/DatePicker/DatePickerMenu/DatePickerMenu.tsx (hunk 21 above)
[the remark this answers is not in the document]
Comment (TheSonOfThomp) [reply packages/date-picker/src/DatePicker/DatePickerMenu/DatePickerMenu.tsx:43]: I might add some additional tests to ensure this is correct though

File: packages/date-picker/src/DatePicker/DatePickerMenu/DatePickerMenu.tsx (hunk 20 above)
Comment (TheSonOfThomp) [reply packages/date-picker/src/DatePicker/DatePickerMenu/DatePickerMenu.tsx:145]: Agree, but that will take some time to abstract correctly. I'd suggest making a ticket for that and coming back to this later

**Comment by bruugey** [comment]
- If you change the date via the segment inputs then you are not able to tab to the CalendarCell components. However, if you just open the menu without having touched the Segment inputs you are able to focus on the CalendarCell components.

**Comment by TheSonOfThomp** [comment]
> Arrow key-ing on year in the demo gives me a jump from 1970 => 2038

This is as-designed since those are the min & max years by default. Date segments will roll-over from min to max (i.e. month 1 to month 12)

File: packages/date-picker/src/DatePicker/SingleDateContext/SingleDateContext.tsx [hunk 24]
```diff
@@ -0,0 +1,187 @@
...
+  children,
+  value,
+  setValue: _setValue,
+  handleValidation,
+}: PropsWithChildren<SingleDateProviderProps>) => {
+  const refs = useDateRangeComponentRefs();
+  const { isOpen, setOpen } = useDatePickerContext();
+  const prevValue = usePrevious(value);
+
+  const today = useMemo(() => setToUTCMidnight(new Date(Date.now())), []);
```
Comment (bruugey) [inline packages/date-picker/src/DatePicker/SingleDateContext/SingleDateContext.tsx:46]: we might not want to memoize "today" as one interaction with the datepicker could span a few minutes (days - if 11:59 -> 12:02), and this could easily get stale.

File: packages/date-picker/src/DatePicker/DatePickerComponent/DatePickerComponent.tsx [hunk 25]
```diff
@@ -0,0 +1,116 @@
...
+    menuTriggerEvent,
+    handleValidation,
+    getHighlightedCell,
+  } = useSingleDateContext();
+
+  const formFieldRef = useForwardedRef(fwdRef, null);
+  const menuRef = useRef<HTMLDivElement>(null);
+
+  useBackdropClick(closeMenu, [formFieldRef, menuRef], isOpen);
```
Comment (bruugey) [inline packages/date-picker/src/DatePicker/DatePickerComponent/DatePickerComponent.tsx:37]: should not fire when a select is open

File: packages/date-picker/src/shared/components/DateInput/DateInputSegment/DateInputSegment.tsx [hunk 26]
```diff
@@ -0,0 +1,153 @@
...
+      }
+
+      onKeyDown?.(e);
+    };
+
+    // Note: Using a text input with pattern attribute due to Firefox
+    // stripping leading zeros on number inputs - Thanks @matt-d-rat
+    // Number inputs also don't support the `selectionStart`/`End` API
+    return (
+      <input
```
Comment (bruugey) [inline packages/date-picker/src/shared/components/DateInput/DateInputSegment/DateInputSegment.tsx:122]: `autocomplete={false}`

File: packages/date-picker/src/DatePicker/SingleDateContext/SingleDateContext.tsx (hunk 24 above)
Comment (TheSonOfThomp) [reply packages/date-picker/src/DatePicker/SingleDateContext/SingleDateContext.tsx:46]: Shouldn't be too expensive. I'd want to add tests for this edge case though

File: packages/date-picker/src/DatePicker/DatePickerComponent/DatePickerComponent.tsx (hunk 25 above)
Comment (TheSonOfThomp) [reply packages/date-picker/src/DatePicker/DatePickerComponent/DatePickerComponent.tsx:37]: I believe we `stopPropagation` within the menu element. Are you seeing a bug?

**Comment by TheSonOfThomp** [comment]
> * If you change the date via the segment inputs then you are not able to tab to the CalendarCell components. However, if you just open the menu without having touched the Segment inputs you are able to focus on the CalendarCell components.

Ticket: https://jira.mongodb.org/browse/LG-3810

File: packages/date-picker/src/DatePicker/SingleDateContext/SingleDateContext.tsx (hunk 24 above)
Comment (TheSonOfThomp) [reply packages/date-picker/src/DatePicker/SingleDateContext/SingleDateContext.tsx:46]: Ticket: [Remove memoization of today](https://jira.mongodb.org/browse/LG-3812)

File: packages/date-picker/src/DatePicker/DatePickerComponent/DatePickerComponent.tsx (hunk 25 above)
Comment (bruugey) [reply packages/date-picker/src/DatePicker/DatePickerComponent/DatePickerComponent.tsx:37]: Yes. if you click outside an open date picker when a select is open both close

File: packages/date-picker/src/shared/components/DatePickerContext/DatePickerContext.tsx [hunk 27]
```diff
@@ -0,0 +1,57 @@
...
+  defaultDatePickerContext,
+  getContextProps,
+} from './DatePickerContext.utils';
+
+/** Create the DatePickerContext */
+export const DatePickerContext = createContext<DatePickerContextProps>(
+  defaultDatePickerContext,
+);
+
+// TODO: Consider renaming this to `SharedDatePickerContext`,
```
[the remark this answers is not in the document]
Comment (TheSonOfThomp) [reply packages/date-picker/src/shared/components/DatePickerContext/DatePickerContext.tsx:20]: Yeah, I think that name is clearer. Just super low priority

File: packages/date-picker/src/DatePicker/DatePickerComponent/DatePickerComponent.tsx [hunk 28]
```diff
@@ -0,0 +1,116 @@
...
+        refEl={formFieldRef}
+        onKeyDown={handleDatePickerKeyDown}
+        onTransitionEnd={handleMenuTransitionEntered}
+        onExited={handleMenuTransitionExited}
+      />
+    </>
+  );
+});
+
+DatePickerComponent.displayName = 'DatePickerContents';
```
Comment (DesignerDave) [inline packages/date-picker/src/DatePicker/DatePickerComponent/DatePickerComponent.tsx:116]: Mismatch of displayName and the component's name.

**Review by DesignerDave — COMMENTED** [review]
Some comments – leaving em as non-blocking but would be good to resolve! 😄

File: packages/date-picker/src/DatePicker/DatePicker.tsx [hunk 29]
```diff
@@ -0,0 +1,70 @@
...
+import { pickAndOmit } from '../shared/utils';
+
+import { DatePickerProps } from './DatePicker.types';
+import { DatePickerComponent } from './DatePickerComponent';
+import { SingleDateProvider } from './SingleDateContext';
+
+/**
+ * LeafyGreen Date Picker component
+ */
+export const DatePicker = forwardRef<HTMLDivElement, DatePickerProps>(
```
[the remark this answers is not in the document]
Comment (TheSonOfThomp) [reply packages/date-picker/src/DatePicker/DatePicker.tsx:23]: There isn't a single `InputElement` to attach to in this case—there are 3 separate segments for each d/m/y value.

As for the fwdRef levels, we're only passing it down one level before we would need to access it (DP > DPComponent)

File: packages/date-picker/src/DatePicker/DatePickerComponent/DatePickerComponent.tsx [hunk 30]
```diff
@@ -0,0 +1,116 @@
...
+import { keyMap } from '@leafygreen-ui/lib';
+
+import { useDatePickerContext } from '../../shared/components/DatePickerContext';
+import { DatePickerInput } from '../DatePickerInput';
+import { DatePickerMenu } from '../DatePickerMenu';
+import { useSingleDateContext } from '../SingleDateContext';
+
+import { DatePickerComponentProps } from './DatePickerComponent.types';
+
+export const DatePickerComponent = forwardRef<
```
[the remark this answers is not in the document]
Comment (TheSonOfThomp) [reply packages/date-picker/src/DatePicker/DatePickerComponent/DatePickerComponent.tsx:20]: Considering rolling back to `DatePickerContents`. Thoughts @shaneeza @bruugey ?

File: packages/date-picker/src/DatePicker/DatePicker.tsx [hunk 31]
```diff
@@ -0,0 +1,70 @@
...
+    );
+
+    return (
+      <DatePickerProvider {...contextProps}>
+        <SingleDateProvider
+          value={value}
+          setValue={setValue}
+          handleValidation={handleValidation}
+        >
+          <LeafyGreenProvider
```
[the remark this answers is not in the document]
Comment (TheSonOfThomp) [reply packages/date-picker/src/DatePicker/DatePicker.tsx:56]: It's possible to pass in a `darkMode` or `baseFontSize` prop into the component that doesn't match the outer context. In this case, we want to establish a new LGContext to avoid passing down the modified `darkMode`/`baseFontSize` value down multiple levels

File: packages/date-picker/src/DatePicker/DatePickerComponent/DatePickerComponent.tsx (hunk 30 above)
[the remark this answers is not in the document]
Comment (bruugey) [reply packages/date-picker/src/DatePicker/DatePickerComponent/DatePickerComponent.tsx:20]: I support it

File: packages/date-picker/src/DatePicker/DatePicker.tsx (hunk 31 above)
[the remark this answers is not in the document]
Comment (DesignerDave) [reply packages/date-picker/src/DatePicker/DatePicker.tsx:56]: Got it! Missed that we'd need nested components to inherit the local override.

File: packages/date-picker/src/DatePicker/DatePickerComponent/DatePickerComponent.tsx (hunk 25 above)
Comment (TheSonOfThomp) [reply packages/date-picker/src/DatePicker/DatePickerComponent/DatePickerComponent.tsx:37]: @bruugey Can you open a ticket for this if you haven't already ?

File: packages/date-picker/src/DatePicker/DatePickerComponent/DatePickerComponent.tsx [hunk 32]
```diff
@@ -0,0 +1,116 @@
...
+
+  const formFieldRef = useForwardedRef(fwdRef, null);
+  const menuRef = useRef<HTMLDivElement>(null);
+
+  useBackdropClick(closeMenu, [formFieldRef, menuRef], isOpen);
+
+  /** This listens to when the disabled prop changes to true and closes the menu */
+  useEffect(() => {
+    if (disabled) {
+      closeMenu();
```
[the remark this answers is not in the document]
Comment (shaneeza) [reply packages/date-picker/src/DatePicker/DatePickerComponent/DatePickerComponent.tsx:42]: [Renders multiple times when disabled](https://jira.mongodb.org/browse/LG-3809) related to this

File: packages/date-picker/src/DatePicker/DatePicker.tsx (hunk 29 above)
[the remark this answers is not in the document]
Comment (DesignerDave) [reply packages/date-picker/src/DatePicker/DatePicker.tsx:23]: Let's keep this desire from implementers in mind, but in the spirit of our conversation today I'm resolving this one.

File: packages/date-picker/src/DatePicker/DatePickerComponent/DatePickerComponent.tsx (hunk 30 above)
[the remark this answers is not in the document]
Comment (DesignerDave) [reply packages/date-picker/src/DatePicker/DatePickerComponent/DatePickerComponent.tsx:20]: Cool with that too.

File: packages/date-picker/src/DatePicker/DatePickerComponent/DatePickerComponent.tsx (hunk 25 above)
Comment (bruugey) [reply packages/date-picker/src/DatePicker/DatePickerComponent/DatePickerComponent.tsx:37]: https://jira.mongodb.org/browse/LG-3836

File: packages/date-picker/src/DatePicker/DatePickerComponent/DatePickerComponent.tsx (hunk 32 above)
[the remark this answers is not in the document]
Comment (bruugey) [reply packages/date-picker/src/DatePicker/DatePickerComponent/DatePickerComponent.tsx:42]: https://jira.mongodb.org/browse/LG-3837

File: packages/date-picker/src/shared/components/DateInput/DateInputBox/DateInputBox.tsx (hunk 10 above)
Comment (shaneeza) [reply packages/date-picker/src/shared/components/DateInput/DateInputBox/DateInputBox.tsx]: I also thought it was `if`, maybe you should add more comments on this?

**Review by bruugey — APPROVED** [review]
